File: isobuilder/dol.py
import struct
from binaryfile import BinaryFile
from isofile import IsoFile

class DOL(IsoFile):
    """DOL file"""

    def __init__(self, path:str, offset:int=0, size:int=0, file:BinaryFile=None, isSystem=False):
        super().__init__(path, isDir=False, offset=offset, size=size, file=file, isSystem=isSystem)
        self.textSections = []
        self.dataSections = []
        self.bssAddr      = 0
        self.bssSize      = 0
        self.entryPoint   = 0

        if file is not None:
            self.readFile(file, offset)


    def readFile(self, file:(str,BinaryFile), offset:int=0):
        """Read DOL file."""
        if type(file) is str:
            file = BinaryFile(file, 'rb', offset=offset)
        else: file.seek(offset)

        self.textSections = []
        self.dataSections = []

        for i in range( 7): # text section offsets
            self.textSections.append({'offset': file.readu32()})
        for i in range(11): # data section offsets
            self.dataSections.append({'offset': file.readu32()})
        for i in range( 7): # text section addrs
            self.textSections[i]['addr'] = file.readu32()
        for i in range(11): # data section addrs
            self.dataSections[i]['addr'] = file.readu32()
        for i in range( 7): # text section sizes
            self.textSections[i]['size'] = file.readu32()
        for i in range(11): # data section sizes
            self.dataSections[i]['size'] = file.readu32()
        self.bssAddr, self.bssSize, self.entryPoint = file.readu32(3)

        self.size = max(map(
            lambda s: s['offset']+s['size'],
            self.textSections + self.dataSections))

        return self


    def dump(self):
        """Dump to console."""
        print(self.path)
        print("  Section | Offset | MemAddr  | Size")
        for i, sec in enumerate(self.textSections):
            print("  text%-2d  | %06X | %08X | %06X" % (i,
                sec['offset'], sec['addr'], sec['size'] ))
        for i, sec in enumerate(self.dataSections):
            print("  data%-2d  | %06X | %08X | %06X" % (i,
                sec['offset'], sec['addr'], sec['size'] ))
        print("  bss     | ------ | %08X | %06X" % (
            self.bssAddr, self.bssSize))
        print("  entry point: 0x%08X" % self.entryPoint)

File: browser/r/js/types/iso/dol.js
import { hex } from '../../Util.js';
import IsoFile from './isofile.js';

const NUM_TEXT_SECTIONS = 7;
const NUM_DATA_SECTIONS = 11;

export default class DOL extends IsoFile {
    constructor(path, offset=0, size=0, buffer=null, isSystem=false) {
        super(path, false, offset, size, buffer, offset, null, isSystem);
        this.textSections = [];
        this.dataSections = [];
        this.bssAddr      = 0;
        this.bssSize      = 0;
        this.entryPoint   = 0;
        if(buffer != null) this.readBuffer(buffer, offset);
    }

    addrToSection(addr) {
        for(let section of this.textSections.concat(this.dataSections)) {
            if(addr >= section.addr && addr < section.addr+section.size) {
                return section;
            }
        }
        return null;
    }

    addrToOffset(addr) {
        const section = this.addrToSection(addr);
        if(!section) return null;
        return (addr - section.addr) + section.offset;
    }

    readBuffer(buffer, offset) {
        if(buffer.buffer) {
            //if given a typed array, get the underlying buffer.
            offset += buffer.byteOffset;
            buffer  = buffer.buffer;
        }
        const view = new DataView(buffer, offset);
        offset = 0;
        for(let i=0; i<NUM_TEXT_SECTIONS; i++) {
            this.textSections.push({offset:view.getUint32(offset)});
            offset += 4;
        }
        for(let i=0; i<NUM_DATA_SECTIONS; i++) {
            this.dataSections.push({offset:view.getUint32(offset)});
            offset += 4;
        }
        for(let i=0; i<NUM_TEXT_SECTIONS; i++) {
            this.textSections[i].addr = view.getUint32(offset);
            offset += 4;
        }
        for(let i=0; i<NUM_DATA_SECTIONS; i++) {
            this.dataSections[i].addr = view.getUint32(offset);
            offset += 4;
        }
        for(let i=0; i<NUM_TEXT_SECTIONS; i++) {
            this.textSections[i].size = view.getUint32(offset);
            offset += 4;
        }
        for(let i=0; i<NUM_DATA_SECTIONS; i++) {
            this.dataSections[i].size = view.getUint32(offset);
            offset += 4;
        }
        this.bssAddr    = view.getUint32(offset);
        this.bssSize    = view.getUint32(offset+4);
        this.entryPoint = view.getUint32(offset+8);
        offset += 12;

        this.size = Math.max(...(this.textSections.concat(this.dataSections))
            .map(it => it.offset + it.size));
        console.log(`DOL size=0x${hex(this.size)}`);
        console.log(`data: ${hex(this.getRawData().getUint32(0x100))}`);

        this._dumpSections();
    }

    _dumpSections() {
        let secs = {};
        for(let i=0; i<NUM_TEXT_SECTIONS; i++) {
            const sec = this.textSections[i];
            secs[`text${i}`] = {
                addr:    hex(sec.addr),
                offset:  hex(sec.offset),
                size:    hex(sec.size),
                end:     hex(sec.addr+sec.size),
                endOffs: hex(sec.offset+sec.size),
            };
        }
        for(let i=0; i<NUM_DATA_SECTIONS; i++) {
            const sec = this.dataSections[i];
            secs[`data${i}`] = {
                addr:    hex(sec.addr),
                offset:  hex(sec.offset),
                size:    hex(sec.size),
                end:     hex(sec.addr+sec.size),
                endOffs: hex(sec.offset+sec.size),
            };
        }
        secs['bss'] = {
            addr:    hex(this.bssAddr),
            offset:  null,
            size:    hex(this.bssSize),
            end:     hex(this.bssAddr+this.bssSize),
            endOffs: null,
        };
        console.table(secs);
    }
}
